### PR TITLE
Conditional random fields

### PR DESCRIPTION
**Pull request by bramvandeneijnden**
extension of random field generator with random field conditioning.

Uses `sklear` as additional requirement for efficient kriging, which is not available efficiently enough in gstools.

### CODE REVIEW

File: docs/contributions.rst [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        """
+
+        if model.project_parameters is None:
+            raise ValueError("Solver settings are not initialised in model.")
+
+        return self.solver_io.create_settings_dictionary(
+            model.project_parameters,
+            Path(mesh_file_name).stem,
+            materials_file_name,
+            model.get_all_model_parts(),
```
Comment (aronnoordam) [inline docs/contributions.rst]: this function and the class above do not exists within the RF repos, can you replace it with an existing function?

File: docs/installation.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+RandomFields installation
+=================
+
+.. _python_stem:
+
+RandomFields Python package
+-------------------
+
+
+To be copied from STEM
```
Comment (aronnoordam) [inline docs/installation.rst:10]:
```suggestion
Work in progress
```

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
 numpy==1.24.2
 gstools==1.5.0
 matplotlib>=3.7.0
+sklearn
```
Comment (aronnoordam) [inline requirements.txt]: missing version, also sklearn is deprecated, i think it should be scikit-learn

File: setup.cfg [hunk 4]
```diff
@@ -17,6 +17,7 @@ install_requires =
     numpy>=1.24
     gstools>=1.5.0
     matplotlib>=3.7.1
+    scikit-learn
```
Comment (aronnoordam) [inline setup.cfg:20]: missing version

File: docs/installation.rst [hunk 5]
```diff
@@ -0,0 +1,10 @@
+RandomFields installation
+=================
+
+.. _python_stem:
```
Comment (aronnoordam) [inline docs/installation.rst]:
```suggestion
.. _python_random_fields:
```

File: random_fields/generate_field.py [hunk 6]
```diff
@@ -4,6 +4,12 @@
 import gstools as gs
```
Comment (aronnoordam) [inline random_fields/generate_field.py:5]:
```suggestion
import numpy.typing as npt
import gstools as gs
```

File: random_fields/generate_field.py [hunk 7]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
         model = self.random_field_model(dim=self.n_dim, var=self.variance, len_scale=scale_fluctuation, angles=self.angle)
         self.random_field = gs.SRF(model, mean=self.mean, seed=self.seed)
         self.random_field(nodes.T)
+
+
+    def set_conditioning_points(self,
+                                points: np.ndarray,
+                                values: np.ndarray,
```
Comment (aronnoordam) [inline random_fields/generate_field.py]:
```suggestion
                                points: npt.NDArray[np.float64],
                                values: npt.NDArray[np.float64],
```

Comment (aronnoordam) [reply random_fields/generate_field.py]: furthermore, can you maintain the style, as described in the contributions section in the docs? I know the old code is not like that, but for new code, i think its better if its done now

File: random_fields/generate_field.py [hunk 8]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+            The (calibrated) correlation kernel from sklearn (default = None)
+        noise_level: float or array-like
+            normalised variance of the observed data, as a single float or as 
+            an array of the same dimensions as `values` (default = 0.0001)
+
+        """
+        self.noise_level = noise_level
+
+        if max(points.shape) > 2000:
+            print('too many conditioning points!')
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: why is this? And can you described it in a comment

File: random_fields/generate_field.py [hunk 9]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        kernel: sklearn kernel
+            The (calibrated) correlation kernel from sklearn (default = None)
+        noise_level: float or array-like
+            normalised variance of the observed data, as a single float or as 
+            an array of the same dimensions as `values` (default = 0.0001)
+
+        """
+        self.noise_level = noise_level
+
+        if max(points.shape) > 2000:
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: either check for points.shape[0] or points.shape[1].  because i think they are always n by ndim or ndim by n

File: random_fields/generate_field.py (hunk 8 above)
Comment (aronnoordam) [inline random_fields/generate_field.py]:
```suggestion
            raise Exception(f'Too many conditioning points! There are {max(points.shape)} points, while the maximum allowed amount is {max_points}')
```

File: random_fields/generate_field.py (hunk 9 above)
Comment (aronnoordam) [inline random_fields/generate_field.py]:
```suggestion
        max_points = 2000
        if max(points.shape) > max_points:
```

File: random_fields/generate_field.py [hunk 10]
```diff
@@ -4,6 +4,12 @@
 import gstools as gs
 
 
+from sklearn.gaussian_process import GaussianProcessRegressor as GPR
+from sklearn.gaussian_process.kernels import RBF, \
+        ConstantKernel as C, \
+        WhiteKernel as W, Matern
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: can you just keep the names ConstantKernel and WhiteKernel, else later in the code its hard to read what is going on

File: requirements.txt (hunk 3 above)
Comment (aronnoordam) [reply requirements.txt]: also the dependency should be added to requirements_dev.txt

File: random_fields/generate_field.py [hunk 11]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+
+        # apply the anisotropy to the other dimensions
+        mask = np.arange(len(scale_fluctuation)) != self.v_dim
+        scale_fluctuation[mask] = scale_fluctuation[mask] * self.anisotropy
+
+        #
+        # correct the length scales between libraries
+        if self.random_field_model_name == 'Gaussian':
+            ls_sklearn  = scale_fluctuation * 2/np.pi
+            self.KrigingKernel = W(noise_level) + RBF(length_scale = ls_sklearn)
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: variables should be written in snake_case. Also KrigingKernel is only used in this function, so i dont think the "self" is needed. so: kriging_kernel

File: random_fields/generate_field.py [hunk 12]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+            ls_sklearn  = scale_fluctuation * 2/np.pi
+            self.KrigingKernel = W(noise_level) + RBF(length_scale = ls_sklearn)    
+        elif self.random_field_model_name == 'Exponential':
+            ls_sklearn  = scale_fluctuation / np.sqrt(1/2)
+            self.KrigingKernel = W(noise_level) + Matern(length_scale = ls_sklearn, nu = 0.5)    
+        elif self.random_field_model_name == 'Matern':
+            ls_sklearn  = scale_fluctuation / np.sqrt(1/2)
+            self.KrigingKernel = W(noise_level) + Matern(length_scale = ls_sklearn)    
+
+        self.GaussianProcess = GPR(kernel = self.KrigingKernel)
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: gaussian_process

File: random_fields/generate_field.py [hunk 13]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        self.GaussianProcess.optimizer = None
+        self.GaussianProcess.fit(points,(values - self.mean) / np.sqrt(self.variance)  )
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: why do you normalise?

File: random_fields/generate_field.py [hunk 14]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        self.conditioning_points = points
+        self.conditioning_values = values
+        
+        self.kriging_mean = None
+        self.kriging_std = None
+
+
+    def generate_conditioned(self, nodes: np.ndarray) -> None:
```
Comment (aronnoordam) [inline random_fields/generate_field.py]:
```suggestion
    def generate_conditioned(self, nodes: npt.NDArray[np.float64]) -> None:
```

File: random_fields/generate_field.py [hunk 15]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+            self.KrigingKernel = W(noise_level) + Matern(length_scale = ls_sklearn)    
+
+        self.GaussianProcess = GPR(kernel = self.KrigingKernel)
+        self.GaussianProcess.optimizer = None
+        self.GaussianProcess.fit(points,(values - self.mean) / np.sqrt(self.variance)  )
+        self.conditioning_points = points
+        self.conditioning_values = values
+        
+        self.kriging_mean = None
+        self.kriging_std = None
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: You create multiple "self.{variable_name}"s here. These variables should be inititalised under, the __init__ function , such that all the member variables are know, and can be described from the beginning

File: random_fields/generate_field.py [hunk 16]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        # create kriged mean field
+        z_kriged_field,std_kriged_field  = self.GaussianProcess.predict(nodes,return_std = True)
+
+        std_kriged_field = np.sqrt(std_kriged_field**2 - self.noise_level)
+
+        std_kriged_field *= np.sqrt(self.variance)
+        self.kriging_mean = self.mean + np.sqrt(self.variance) * z_kriged_field
+        #
+        # create single random field at nodes and conditioning points
+        nodes_cpoints = np.vstack([nodes,self.conditioning_points])
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: can you write out the variable names a bit more, e.g. nodes_cond_points. Also for example z_cond_rf_nodes instead of z_crf_nodes. This makes the code a bit more understandable

File: random_fields/generate_field.py [hunk 17]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        self.generate(nodes_cpoints)
+
+        z_crf_nodes = (self.random_field.field[:nodes.shape[0]] - self.mean) / np.sqrt(self.variance)
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: can you describe what you are doing

File: random_fields/generate_field.py [hunk 18]
```diff
@@ -94,3 +102,108 @@ def generate(self, nodes: np.ndarray) -> None:
...
+
+        z_crf_nodes = (self.random_field.field[:nodes.shape[0]] - self.mean) / np.sqrt(self.variance) 
+        z_rf_cpoints = (self.random_field.field[nodes.shape[0]:] - self.mean) / np.sqrt(self.variance) 
+        
+        np.random.seed(self.seed)
+        z_rf_cpoints += np.sqrt(self.noise_level) * np.random.normal(size = z_rf_cpoints.shape)
+        
+        #
+        # create kriged mean field of random field 
+        GPrf = GPR(kernel = self.GaussianProcess.kernel_)
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: snake_case

**Review by aronnoordam — COMMENTED** [review]
few comments about style and some stuff what is unclear for me.

Furthermore can you add tests for the added functionalities?

File: random_fields/generate_field.py [hunk 19]
```diff
@@ -91,6 +145,128 @@ def generate(self, nodes: np.ndarray) -> None:
...
+        z_cond_RF_nodes = (self.random_field.field[:nodes.shape[0]] - self.mean) / np.sqrt(self.variance) 
+        z_RF_cond_points = (self.random_field.field[nodes.shape[0]:] - self.mean) / np.sqrt(self.variance) 
+        
+        # add the noise to the conditioning points
+        np.random.seed(self.seed)
+        z_RF_cond_points += np.sqrt(self.noise_level) * np.random.normal(size = z_RF_cond_points.shape)
+        
+        #
+        # create kriged mean field of random field 
+        GP_RF = GPR(kernel = self.GaussianProcess.kernel_)
```
Comment (aronnoordam) [inline random_fields/generate_field.py]: lower case

File: tests/test_random_field.py [hunk 20]
```diff
@@ -86,14 +86,70 @@ def test_distribution_RF_unstruc():
...
+
+    # evaluate and test maximum difference 
+    max_error_mean = np.max(np.abs(kriging_mean_ref - rf.kriging_mean))
+    max_error_std = np.max(np.abs(kriging_std_ref - rf.kriging_std))
+
+    np.testing.assert_array_almost_equal(max_error_mean, 0., decimal=4)
+    np.testing.assert_array_almost_equal(max_error_std, 0., decimal=4)
+
+
+test_conditioned_RF_3D_mean_variance()
```
Comment (aronnoordam) [inline tests/test_random_field.py]:
```suggestion
```

File: tests/test_random_field.py [hunk 21]
```diff
@@ -86,14 +86,70 @@ def test_distribution_RF_unstruc():
...
+    # generate and plot conditioned random field model
+    rf.generate_conditioned(np.array([x, y,z]).T)
+
+    rf.conditioned_random_field
+    rf.kriging_mean
+    rf.kriging_std
+
+    ## to regenerate new reference solutions
+    # np.savetxt('./data/kriging_mean_3D.txt',rf.kriging_mean)
+    # np.savetxt('./data/kriging_std_3D.txt',rf.kriging_std)
```
Comment (aronnoordam) [inline tests/test_random_field.py]:
```suggestion
```

File: docs/contributions.rst (hunk 1 above)
Comment (bramvandeneijnden) [reply docs/contributions.rst]: updated with existing function

File: requirements.txt (hunk 3 above)
Comment (bramvandeneijnden) [reply requirements.txt]: scikit-learn corrected ==1.5.1

I tried to set matplotlib version to the one used in the STEM package (3.7.4), but this breaks one of the tests on the plotting utils...

File: setup.cfg (hunk 4 above)
Comment (bramvandeneijnden) [reply setup.cfg:20]: fixed

File: random_fields/generate_field.py (hunk 14 above)
Comment (bramvandeneijnden) [reply random_fields/generate_field.py]: corrected, but hot to properly document arrays in the docstrings? see my attempt in the docstrings

File: random_fields/generate_field.py [hunk 22]
```diff
@@ -28,26 +28,25 @@ class RandomFields():
...
+        - variance (float): unconditioned variance of the random field
+        - vertical_scale_fluctuation (float): 
+        - anisotropy (list): ratio between horizontal scales and vertical scale of fluctuation, shape (`self.n_dims` - 1)
+        - angle (list): rotation angles of the principal directions of the scales of fluctuation relative to the vertical, shape (`self.n_dims` - 1)
+        - v_dim (int): dimesion number corresponding to the vertical scale of fluctuation
+        - random_field (:class:gstools.SRF) random field generator 
+        - z_kriged_field (array): standard-normal kriged random field
+        - gaussian_process (:class:sklearn.gaussian_process.GaussianProcessRegressor)
+        - conditioning_points (array): coordinates of the conditioning points, shape (:,`self.n_dims`)
+        - conditioning_values (array): values of the conditioning points, shape (:,)
```
Comment (bramvandeneijnden) [inline random_fields/generate_field.py:46]: How to document arrays in the doc strings? I've seen "Sequence(float)" here and there... 

and what about shapes and defaults?
